modules/caddytls/internalissuer.go

```diff
@@ -0,0 +1,199 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package caddytls
+
+import (
+	"bytes"
+	"context"
+	"crypto"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"time"
+
+	"github.com/caddyserver/caddy/v2"
+	"github.com/caddyserver/caddy/v2/modules/caddypki"
+	"github.com/caddyserver/certmagic"
+	"github.com/smallstep/certificates/authority"
+	"github.com/smallstep/certificates/authority/provisioner"
+	"github.com/smallstep/cli/crypto/x509util"
+)
+
+func init() {
+	caddy.RegisterModule(InternalIssuer{})
+}
+
+// InternalIssuer is a certificate issuer that generates
+// certificates internally using a locally-configured
+// CA which can be customized using the `pki` app.
+type InternalIssuer struct {
+	// The ID of the CA to use for signing. The default
+	// CA ID is "local". The CA can be configured with the
+	// `pki` app.
+	CA string `json:"ca,omitempty"`
+
+	// The validity period of certificates.
+	Lifetime caddy.Duration `json:"lifetime,omitempty"`
+
+	// If true, the root will be the issuer instead of
+	// the intermediate. This is not recommended and should
+	// only be used when devices/clients do not properly
+	// validate certificate chains.
+	SignWithRoot bool `json:"sign_with_root,omitempty"`
+
+	ca *caddypki.CA
+}
+
+// CaddyModule returns the Caddy module information.
+func (InternalIssuer) CaddyModule() caddy.ModuleInfo {
+	return caddy.ModuleInfo{
+		ID:  "tls.issuance.internal",
+		New: func() caddy.Module { return new(InternalIssuer) },
+	}
+}
+
+// Provision sets up the issuer.
+func (li *InternalIssuer) Provision(ctx caddy.Context) error {
+	// get a reference to the configured CA
+	appModule, err := ctx.App("pki")
+	if err != nil {
+		return err
+	}
+	pkiApp := appModule.(*caddypki.PKI)
+	if li.CA == "" {
+		li.CA = defaultInternalCAName
+	}
+	ca, ok := pkiApp.CAs[li.CA]
+	if !ok {
+		return fmt.Errorf("no certificate authority configured with id: %s", li.CA)
+	}
+	li.ca = ca
+
+	// set any other default values
+	if li.Lifetime == 0 {
+		li.Lifetime = caddy.Duration(defaultInternalCertLifetime)
+	}
+
+	return nil
+}
+
+// IssuerKey returns the unique issuer key for the
+// confgured CA endpoint.
+func (li InternalIssuer) IssuerKey() string {
+	return li.ca.ID()
+}
+
+// Issue issues a certificate to satisfy the CSR.
+func (li InternalIssuer) Issue(ctx context.Context, csr *x509.CertificateRequest) (*certmagic.IssuedCertificate, error) {
+	// prepare the signing authority
+	// TODO: eliminate placeholders / needless values
+	cfg := &authority.Config{
+		Address:          "placeholder_Address:1",
+		Root:             []string{"placeholder_Root"},
+		IntermediateCert: "placeholder_IntermediateCert",
+		IntermediateKey:  "placeholder_IntermediateKey",
+		DNSNames:         []string{"placeholder_DNSNames"},
+		AuthorityConfig: &authority.AuthConfig{
+			Provisioners: provisioner.List{},
+		},
+	}
```

Comment:
@mmalone @maraino Do you know if there's any chance we can drop these placeholders/meaningless values?